prover/src/constraints/evaluator/logup_gkr.rs

```diff
@@ -94,16 +115,44 @@ impl<E: FieldElement> LagrangeKernelConstraintsBatchEvaluator<E> {
 
                 // combine boundary constraints
                 {
-                    let boundary_numerator =
-                        self.lagrange_kernel_constraints.boundary.evaluate_numerator_at(&frame);
+                    let boundary_numerator = self
+                        .lagrange_kernel_constraints
+                        .boundary
+                        .evaluate_numerator_at(&lagrange_frame);
 
                     combined_evaluations += boundary_numerator * boundary_divisors_inv[step];
                 }
 
                 combined_evaluations
             };
 
-            combined_evaluations_acc[step] += combined_evaluations;
+            // compute and combine the transition constraints for the s-column.
+            // The s-column implements the cohomological sum-check argument of [1] and
+            // the constraint we enfore is exactly Eq (4) in Lemma 1 in [1].
+            //
+            // [1]: https://eprint.iacr.org/2021/930
+            let s_col_combined_evaluation = {
+                trace.read_main_trace_frame_into(step << lde_shift, &mut main_frame);
+                trace.read_aux_trace_frame_into(step << lde_shift, &mut aux_frame);
+
+                let l_cur = aux_frame.current()[l_col_idx];
+                let s_cur = aux_frame.current()[s_col_idx];
+                let s_nxt = aux_frame.next()[s_col_idx];
+
+                evaluator.build_query(&main_frame, &[], &mut query);
+                let batched_query = self.gkr_data.compute_batched_query(&query);
+
+                let rhs = s_cur - mean + batched_query * l_cur;
+                let lhs = s_nxt;
+
+                (rhs - lhs)
+                    * self
+                        .s_col_composition_coefficient
+                        .mul_base(s_col_constraint_divisor[step % (domain.trace_to_ce_blowup())])
```

Comment:
nit: I would define `s_col_constraint_divisor[step % (domain.trace_to_ce_blowup())]` as a separate variable to make this a bi more readable.

Reply:
Updated